aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -515,6 +515,369 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+        out.copy_(at::matrix_power(in, n));
+      }
+    );
+
+    return res;
+  }
+  else {
+    int64_t n = powers.item().template to<int64_t>();
+    return at::matrix_power(matrices, n);
+  }
+}
+
+Tensor operator_1_norm(const Tensor& tensor) {
+  return std::get<0>(tensor.abs().sum(-2).max(-1));
+}
+
+Tensor compute_T1(const Tensor& A) {
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  return I + A;
+}
+
+Tensor compute_T2(const Tensor& A) {
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  const auto A2 = at::matmul(A, A);
+  return I + A + A2 / fact[2];
+}
+
+Tensor compute_T4(const Tensor& A) {
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  const auto A2 = at::matmul(A, A);
+  return I + A + at::matmul(A2, I / fact[2] + A / fact[3] + A2 / fact[4]);
+}
+
+template <typename scalar_t>
+Tensor compute_T8(const Tensor& A) {
+  constexpr scalar_t sqrt_177 = 0.1330413469565007072504e+2;
+  constexpr scalar_t x3 = 2. / 3.;
+  constexpr scalar_t x1 = x3 * ((1. + sqrt_177) / 88.);
+  constexpr scalar_t x2 = x3 * ((1. + sqrt_177) / 352.);
+  constexpr scalar_t x4 = (-271. + 29. * sqrt_177) / (315. * x3);
+  constexpr scalar_t x5 = (-11. + 11. * sqrt_177) / (1260. * x3);
+  constexpr scalar_t x6 = (-99. + 11. * sqrt_177) / (5040. * x3);
+  constexpr scalar_t x7 = (89. - sqrt_177) / (5040. * x3);
+  constexpr scalar_t y2 = (857. - 58. * sqrt_177) / 630.;
+
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  const auto A2 = at::matmul(A, A);
+  const auto A4 = at::matmul(A2, x1 * A + x2 * A2);
+  const auto A8 = at::matmul(x3 * A2 + A4,
+    x4 * I + x5 * A + x6 * A2 + x7 * A4);
+  return I + A + y2 * A2 + A8;
+}
+
+template <typename scalar_t>
+Tensor compute_T12(const Tensor& A) {
+  constexpr int num_prods = 4;
+  constexpr array2d<scalar_t, num_prods, num_prods> b = {{
+    {
+      9.0198e-16,
+      0.46932117595418237389,
+      -0.20099424927047284052,
+      -0.04623946134063071740
+    },
+    {
+      5.31597895759871264183,
+      1.19926790417132231573,
+      0.01179296240992997031,
+      0.01108844528519167989
+    },
+    {
+      0.18188869982170434744,
+      0.05502798439925399070,
+      0.09351590770535414968,
+      0.00610700528898058230
+    },
+    {
+      -2.0861320e-13,
+      -0.13181061013830184015,
+      -0.02027855540589259079,
+      -0.00675951846863086359
+    }
+  }};
+
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  const auto A2 = at::matmul(A, A);
+  const auto A3 = at::matmul(A2, A);
+  std::array<
+    std::reference_wrapper<const Tensor>,
+    num_prods> As = {I, A, A2, A3};
+
+  std::array<Tensor, num_prods> Bs;
+  for (int i = 0; i < num_prods; ++i) {
+    Bs[i] = at::zeros(A.sizes(), A.options());
+  }
+
+  for (int i = 0; i < num_prods; ++i) {
+    for (int j = 0; j < num_prods; ++j) {
+      Bs[i] += b[i][j] * As[j];
+    }
+  }
+
+  const auto A6 = Bs[2] + at::matmul(Bs[3], Bs[3]);
```

Comment:
`addmm` would be beneficial here.

Reply:
Yes, but the documentation on `addmm` does not say anything about supporting batch dimensions, while `matmul` and `+` does support batching. I would rather leave `comput_T{i}` the way they are, as they support batching.

Reply:
Grouping by norm is going to be super important for CUDA too...

Reply:
We can do that and use all degree approximations even for CUDA to safe on matrix multiply during the "squaring" procedure. This will require syncing with the host not read out the norms. In the current implementation (not yet pushed), I still synchronize once to read out the number of steps the "squaring" procedure needs to be done per matrix.
Def. "squaring" is a successive x = x @ x.

Reply:
Another benefit is memory. Using the highest approximation formula requires at least 10x time memory than the size of the input...